src/util/timer.cpp

```diff
@@ -1,246 +1,113 @@
-  }
-  return {{"data", arr}, {"children", json_chldrn}};
-#else
-  return {};
-#endif
-}
+      return obj;
+    }  
+  };
 
-// otto::util::timer::ScopedTimer /////////////////////////////////////////////
+  static std::mutex mutex;
+  static std::map<std::string, ThreadTimer> thread_timers;
+
+  static std::string get_thread_id() {
+    static thread_local std::string thread_id;
+    if (thread_id.empty()) {
+      std::ostringstream s;
+      s << "Thread " << std::this_thread::get_id();
+      thread_id = s.str();
+    }
 
-  ScopedTimer::ScopedTimer(Timer& timer) noexcept
-#if ENABLED
-    : timer (&timer)
-#endif
-  {}
+    return thread_id;
+  }
 
-  ScopedTimer::ScopedTimer(Timer* timer) noexcept
-#if ENABLED
-    : timer (timer)
-#endif
-  {}
+  static ThreadTimer& get_thread_timer() {
+    std::lock_guard lock(mutex);
+    auto thread_id = get_thread_id();
+    auto [iter, created] = thread_timers.emplace(thread_id, thread_id);
 
-  ScopedTimer::ScopedTimer(ScopedTimer&& rhs) noexcept
-#if ENABLED
-    : timer (rhs.timer)
-#endif
-  {
-#if ENABLED
-    rhs.timer = nullptr;
-#endif
+    return iter->second;
   }
 
-  ScopedTimer::~ScopedTimer()
+  ScopedTimer::ScopedTimer(std::string id)
+    : id(id), start(time_point::clock::now())
   {
-#if ENABLED
-    if (timer != nullptr) stop();
-#endif
+    auto& thread_timer = get_thread_timer();
+    parent = thread_timer.active_timer;
+    thread_timer.active_timer = this;
   }
 
-// otto::util::timer free functions ///////////////////////////////////////////
-
-#if ENABLED
-  struct TimerStack {
-    Timer root;
-    std::vector<Timer*> stack;
-
-    TimerStack(const timer_id& name)
-      : root (name),
-        stack {&root}
-    {}
-  };
-
-  /// This owns one vector per thread
-  std::vector<TimerStack> thread_stacks;
-
-  static std::mutex mutex;
+  ScopedTimer::~ScopedTimer() {
+    auto duration = (time_point::clock::now() - start).count();
 
-  static TimerStack* create_timer_stack() {
-    std::ostringstream s;
-    s << "Thread ";
-#if defined (__unix__) || (defined (__APPLE__) && defined (__MACH__))
-    char buf[256];
-    s << pthread_getname_np(pthread_self(), buf, 256);
-#else
-    s << std::this_thread::get_id();
-#endif
-    {
-      std::lock_guard lock (mutex);
-      return &thread_stacks.emplace_back(s.str());
-    }
-  }
+    auto& thread_timer = get_thread_timer();
 
-  // Do not access directly! use the function below
-  thread_local TimerStack* _timer_stack = nullptr;
+    auto& target = (parent == nullptr)
+      ? thread_timer.history
+      : parent->children;
 
-  static auto& timer_stack() noexcept
-  {
-    if (_timer_stack == nullptr) {
-      _timer_stack = create_timer_stack();
-    }
-    return _timer_stack->stack;
+    target.emplace_back(TimerData{id, duration, children});
+    thread_timer.active_timer = parent;
   }
-#else
-  Timer one_timer_to_rule_them_all("");
-#endif
 
-  Timer& find_or_make(timer_id id)
+  ScopedTimer start(std::string id)
```

Comment:
should probably be marked `[[nodiscard]]`

Reply:
I'll give that a try.

Reply:
this is btw related to the whole RAII thing of scoped timers

Reply:
Done.